doc/arify.1
.TH ARIFY 1 "Jan 02, 2023" 0.1.0 "ARIF User Manual"
.
.SH NAME
arify - execute program with input methods
.
.SH SYNOPSIS
.B arify
.RI [ options ]
.I pathname
.RI [ args ]
.
.SH DESCRIPTION
Executes the program referred to by
.IR pathname ,
passing
.I args
as its command-line arguments.
.PP
If the program uses a line editor supported by ARIF,
it will be provided with the ability to interact with input method engines.
The exact behaviour depends on the frontend specified.
.
.SH OPTIONS
.TP
\fB\-e\fR \fIengines\fR
Overrides the
.B ARIFY_ENGINES
environment variable with
.IR engines .
.IP
If multiple
.B \-e
options are provided, the resulting value is a concatenation of each
.IR engines ,
with an ASCII Space character as separator.
.TP
\fB\-f\fR \fIfrontend\fR
Overrides the
.B ARIFY_FRONTEND
environment variable with
.IR frontend .
.TP
\fB\-l\fR \fIlog-file\fR
Overrides the
.B ARIFY_LOG_FILE
environment variable with
.IR log-file .
.TP
\fB\-n\fR \fIpage-size\fR
Overrides the
.B ARIFY_PAGE_SIZE
environment variable with
.IR page-size .
.TP
\fB\-p\fR \fIpreload-library\fR
Path to the preload library (e.g. "/usr/local/lib/libarify.so").
.IP
Default value depends on the specified install directory when building ARIF.
.TP
\fB\-V\fR
Prints the version number of ARIF, and then exits.
.
.SH EXIT STATUS
If an error occurs before
.I pathname
is executed, the program exits with status 1.
Otherwise, the exit status equals to that program's exit status.
.
.SH ENVIRONMENT
.TP
.B ARIFY_FRONTEND
The frontend program to use.
.IP
If the variable is unset or empty, value "readline" will be used.
.IP
To use the Readline frontend, key bindings should be configured.
See the "Readline key bindings" subsection in NOTES for details.
.TP
.B ARIFY_ENGINES
Comma-separated list of input method engines to use.
At least one engine should be specified.
.IP
See the "Engine names" subsection in NOTES for engine name format.
.TP
.B ARIFY_LOG_FILE
Path to the file where error logs (and debug logs, if enabled)
will be written to.
.IP
If the variable is unset or empty, logs will not be printed.
.TP
.B ARIFY_PAGE_SIZE
Maximum number of candidates per page (at most 20).
.IP
If the variable is unset or empty, value 5 will be used.
.SS
Readline-only environments:
.TP
.B ARIFY_RL_NO_AUTO_UNSETENV
By default, when arify is enabled for the first time
(by invoking the
.B arify\-toggle
Readline command),
environment variables related to arify are unset, so that it won't be
initialized in child processes.
.IP
This behavior can be disabled by setting this variable to any non-empty value.
.TP
.B ARIFY_RL_WORD_BREAK_CHARS
List of characters that are supposed to signal a word break.
Overrides
.I rl_completer_word_break_characters
when arify is enabled.
.IP
If the variable is unset or empty, value " \\t\\n" will be used.
.IP
Note that the list should contain the actual characters
(i.e., escape sequences would not work).
.
.SH NOTES
.SS Engine names
Each engine name in the
.B ARIFY_ENGINES
list should comply with the following format:
.PP
.in +4n
.EX
.IR <lib_name> : <symbol_name>
.EE
.in
.PP
The
.I <lib_name>
is the path to the shared library file containing the engine implementation.
This name is used as the first argument to pass to
.BR dlopen (3).
.PP
The
.I <symbol_name>
is the name of the
.B struct arif_engine
where the engine is implemented (see
.BR arif_set_engine (3)).
This name is used as the second argument to pass to
.BR dlsym (3).
.PP
If the two names match the following pattern, a
.I <short_name>
can be used instead of the full name:
.PP
.in +4n
.EX
.IR <libdir> /libarif_ <short_name><shlib_suffix> :arif_ <short_name> _engine
.EE
.in
.PP
Where
.IR <shlib_suffix>
is the common name suffix for shared library files on the current platform
(e.g., ".so" on GNU/Linux).
.SS Readline key bindings
If using the Readline frontend, extra key bindings need to be configured in
.IR \[ti]/.inputrc :
.TP
.B arify\-toggle
Turn on/off input method engine.
.IP
When turned on, Readline's completion functions will be replaced by those
provided by ARIF.
When turned off, the previously saved completion functions will be restored.
.TP
.B arify\-engine\-info
Prints information regarding the currently active engine.
.TP
.B arify\-next\-engine
Switch to the next loaded engine.
If there are no more available engines, switch back to the first engine.
.TP
.B arify\-page\-up
.TQ
.B arify\-page\-down
Switch to previous/next page of candidates, and display them.
.PP
See EXAMPLES for an example configuration.
.
.SH EXAMPLES
.SS Example Readline configuration
The following is a simple
.IR \[ti]/.inputrc
snippet which configures key bindings for arify
when working with
.BR rlexec (1).
.PP
.in +4n
.EX
$if rlexec
    "\\e ":     arify\-toggle
    "\\e\\\\":    arify\-next\-engine
    "\\e[5;3\[ti]": arify\-page\-up      # M\-PgUp
    "\\e[6;3\[ti]": arify\-page\-down    # M\-PgDn
$endif
.EE
.in
.SH BUGS
Theoretically arify should work with most programs with a supported frontend.
However, there are situations where it may not work as expected:
.IP \[bu] 3
When
.B LD_PRELOAD
is not functioning, arify does not work.
.IP \[bu]
When a Readline program does fancy things with the Readline API (e.g.,
change handler functions during user input), arify may also fail to work,
since it may break the internal states arify maintains.
.PP
Meanwhile, adding support for more frontends may not be as simple
(or even possible) as with Readline, due to the way how arify works.
.
.SH COPYRIGHT
Copyright (C) 2023  CismonX <admin@cismon.net>
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU Free Documentation License, Version 1.3
or any later version published by the Free Software Foundation;
with no Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.
.PP
You should have received a copy of the license along with this document.
If not, see
.UR https://www.gnu.org/licenses/fdl-1.3.html
.UE .
.
.SH SEE ALSO
.BR rlexec (1),
.BR arif_set_engine (3),
.BR dlopen (3),
.BR dlsym (3),
.BR readline (3),
.BR ld.so (8)


src/arify.c
/**
 * arif/src/arify.c - ARIF preload library
 * ----
 *
 * Copyright (C) 2023  CismonX <admin@cismon.net>
 *
 * This file is part of ARIF, Another Readline Input Framework.
 *
 * ARIF is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ARIF is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ARIF.  If not, see <https://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#  include "config.h"
#endif

#include "arify.h"

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <dlfcn.h>
#include <fcntl.h>
#include <unistd.h>

#include "arif.h"
#include "arif_common.h"

#ifdef ENABLE_ARIF_READLINE
#  include "arify_rl.h"
#endif

#ifndef ARIF_CTOR
#  error "__attribute__((constructor)) not supported"
#endif

#ifndef ARIFY_MAX_PAGE_SIZE
#  define ARIFY_MAX_PAGE_SIZE  20
#endif

struct arify_engine {
    char const               *lib_name;
    char const               *var_name;
    void                     *handle;
    struct arif_engine const *impl;
    void                     *data;
    struct arify_engine      *next;
};

// Forward declaration start
static int  config_engines   (void);
static int  config_frontend  (void);
static int  config_log_file  (void);
static int  config_page_size (void);
static void finalize         (void);
static void finalize_engines (struct arify_engine *);
static int  init_config      (void);
static void initialize       (void) ARIF_CTOR;
static int  load_engine      (struct arify_engine *);
// Forward declaration end

static struct arify_ctx {
    struct arify_engine         *engines;
    struct arify_engine         *current_engine;
    struct arify_frontend const *frontend;
    void                        *frontend_data;
    struct arif_ctx             *ctx;
    int                          log_fd;
    int                          page_size;
    pid_t                        pid;
} ctx = { .log_fd = -1 };

static int
config_engines (void)
{
    char const *engines_str = xgetenv("ARIFY_ENGINES", NULL);
    if (engines_str == NULL) {
        arify_errf("%s", "no engine specified");
        return -1;
    }

    size_t engines_len = strlen(engines_str) + 1;
    char *names = xmalloc(sizeof(char) * engines_len);
    memcpy(names, engines_str, engines_len);

    struct arify_engine head;
    for (struct arify_engine *engine = &head; ; *(names++) = '\0') {
        char *var_name = strchr(names, ':');
        if (var_name != NULL) {
            *(var_name++) = '\0';
        }
        engine = engine->next = xmalloc(sizeof(*engine));
        *engine = (struct arify_engine) {
            .lib_name = names,
            .var_name = var_name,
        };

        names = strchr(var_name == NULL ? names : var_name, ',');
        if (names == NULL) {
            break;
        }
    }
    ctx.current_engine = ctx.engines = head.next;
    return 0;
}

static int
config_frontend (void)
{
    char const *frontend_str = xgetenv("ARIFY_FRONTEND", "readline");
#ifdef ENABLE_ARIF_READLINE
    if (0 == strcmp("readline", frontend_str)) {
        ctx.frontend = &arify_frontend_readline;
        return 0;
    }
#endif  // defined(ENABLE_ARIF_READLINE)
    arify_errf("unsupported frontend '%s'", frontend_str);
    return -1;
}

static int
config_log_file (void)
{
    char const *log_file_path = xgetenv("ARIFY_LOG_FILE", NULL);
    if (log_file_path != NULL) {
        ctx.log_fd = open(log_file_path, O_WRONLY | O_CREAT | O_APPEND, 0600);
        ctx.pid    = getpid();
    }
    return 0;
}

static int
config_page_size (void)
{
    char const *page_size_str = xgetenv("ARIFY_PAGE_SIZE", "5");
    int page_size = atoi(page_size_str);
    if (page_size < 5 || page_size > ARIFY_MAX_PAGE_SIZE) {
        arify_errf("bad page size, expected [5, %d], got '%s'",
                ARIFY_MAX_PAGE_SIZE, page_size_str);
        return -1;
    }
    ctx.page_size = page_size;
    return 0;
}

static void
finalize (void)
{
    ctx.frontend->finalize(ctx.frontend_data);
    free((char *) ctx.engines->lib_name);
    finalize_engines(ctx.engines);
    arif_ctx_destroy(ctx.ctx);

    arify_debugf("%s", "finalized");
    close(ctx.log_fd);
}

static void
finalize_engines (
    struct arify_engine *engine
) {
    for (struct arify_engine *next; engine != NULL; engine = next) {
        if (engine->impl != NULL) {
            engine->impl->finalize(engine->data);
        }
        if (engine->handle != NULL) {
            dlclose(engine->handle);
        }

        next = engine->next;
        free(engine);
    }
}

static int
init_config (void)
{
    if (0 != config_log_file()) {
        return -1;
    }
    if (0 != config_frontend()) {
        return -1;
    }
    if (0 != config_page_size()) {
        return -1;
    }
    if (0 != config_engines()) {
        return -1;
    }
    return 0;
}

static void
initialize (void)
{
    if (0 != init_config()) {
        if (ctx.log_fd >= 0) {
            close(ctx.log_fd);
        }
        return;
    }
    atexit(finalize);

    struct arif_opts const opts = {
        .page_size = ctx.page_size,
    };
    ctx.ctx = arif_ctx_create(&opts);

    int result = ctx.frontend->init(ctx.ctx, &ctx.frontend_data);
    if (result != 0) {
        arify_errf("frontend init failed with return value %d", result);
        return;
    }

    arify_debugf("%s", "initialized");
}

static int
load_engine (
    struct arify_engine *engine
) {
    char const *lib_name = engine->lib_name;
    char const *var_name = engine->var_name;
    char *lib_tmp = NULL;
    char *var_tmp = NULL;

#define ARIFY_ENGINE_LIB(engine)  \
        ARIF_LIBDIR "/libarif_" engine ARIF_SHLIB_SUFFIX
#define ARIFY_ENGINE_SYM(engine)  "arif_" engine "_engine"

    if (var_name == NULL) {
        size_t lib_name_len = sizeof(ARIFY_ENGINE_LIB("")) + strlen(lib_name);
        lib_tmp = xmalloc(sizeof(char) * lib_name_len);
        sprintf(lib_tmp, ARIFY_ENGINE_LIB("%s"), lib_name);

        size_t var_name_len = sizeof(ARIFY_ENGINE_SYM("")) + strlen(lib_name);
        var_tmp = xmalloc(sizeof(char) * var_name_len);
        sprintf(var_tmp, ARIFY_ENGINE_SYM("%s"), lib_name);

        lib_name = lib_tmp;
        var_name = var_tmp;
    }
    int result = -1;

    void *handle = dlopen(lib_name, RTLD_NOW);
    if (handle == NULL) {
        arify_errf("failed to load engine %s: %s", lib_name, dlerror());
        goto finish;
    }
    engine->handle = handle;

    struct arif_engine const *engine_impl = dlsym(handle, var_name);
    if (engine_impl == NULL) {
        arify_errf("failed to find symbol %s for engine %s: %s",
                var_name, lib_name, dlerror());
        goto finish;
    }

    int init_status = engine_impl->init(NULL, &engine->data);
    if (init_status != 0) {
        arify_errf("engine %s init failed with return value %d",
                lib_name, init_status);
        goto finish;
    }
    engine->impl = engine_impl;
    result = 0;

  finish:
    free(lib_tmp);
    free(var_tmp);
    return result;
}

void
arify_logf (
    char const *fmt,
    char const *level,
    ...
) {
    if (ctx.log_fd < 0) {
        return;
    }
    char log_buf[1024];
    char *buf = log_buf, *end = buf + sizeof(log_buf);

    time_t ts = time(NULL);
    struct tm const *tm = localtime(&ts);
    if (tm == NULL) {
        return;
    }
    buf += strftime(buf, end - buf, "[%T ", tm);
    buf += sprintf(buf, "%ld %s] ", (long) ctx.pid, level);

    va_list args;
    va_start(args, level);
    int len = vsnprintf(buf, end - buf, fmt, args);
    va_end(args);

    if (len >= end - buf) {
        len = end - buf;
        buf[len - 1] = '\n';
    }
    write(ctx.log_fd, log_buf, buf + len - log_buf);
}

struct arif_engine const *
arify_next_engine (
    void **engine_data_ptr
) {
    struct arify_engine      *engine      = ctx.current_engine;
    struct arif_engine const *engine_impl = NULL;
    void                     *engine_data = NULL;
    while (engine != NULL) {
        int result = load_engine(engine);

        engine_impl = engine->impl;
        engine_data = engine->data;
        engine      = engine->next;

        if (result == 0) {
            break;
        }
    }
    ctx.current_engine = engine;

    *engine_data_ptr = engine_data;
    return engine_impl;
}
